xbmc/pvr/channels/PVRChannelProviders.h

```diff
@@ -0,0 +1,121 @@
+/*
+ *  Copyright (C) 2012-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "XBDateTime.h"
+#include "pvr/channels/PVRChannelProvider.h"
+#include "pvr/settings/PVRSettings.h"
+#include "threads/Thread.h"
+
+#include <map>
+#include <memory>
+#include <queue>
+#include <vector>
+
+namespace PVR
+{
+  class CPVRChannelProvider;
+
+  class CPVRChannelProvidersContainer
+  {
+  public:
+    /*!
+     * @brief Add a channel provider to this container or update the provider if already present in this container.
+     * @param The channel provider
+     * @return True, if the update was successful. False, otherwise.
+     */
+    bool UpdateFromClient(const std::shared_ptr<CPVRChannelProvider>& provider);
+
+    /*!
+     * @brief Get the channel provider denoted by given client id and unique client provider id.
+     * @param iClientId The client id.
+     * @param iUniqueId The client provider id.
+     * @return the channel provider if found, null otherwise.
+     */
+    std::shared_ptr<CPVRChannelProvider> GetByClient(int iClientId, int iUniqueId) const;
+
+    std::vector<std::shared_ptr<CPVRChannelProvider>> GetProvidersList() const
+    {
+      return m_providers;
```

Comment:
Mutex protection needed?

Reply:
Here is a general question. If the vector is mutex protected should I instead return a copy?

> Could you add some screen shots?

Done

Reply:
Moved this to protected scope.